src/SINDYPI_fit.jl

```diff
@@ -0,0 +1,457 @@
+Xi = Array{Any}(undef, number_states_guess)
+
+
+ for w = 1:number_states_guess
+    npoly = Polyorder[w]
+    ntrig = Trigorder[w]
+
+    nvar = size(X, 2)
+
+    basis = ones(size(X,1), 1)
+
+    if npoly >=1
+      basis = hcat(basis, X)
+    end
+
+    if npoly >= 2
+        for i in 1:nvar
+            for j in i:nvar
+                basis = hcat(basis, (X[:, i] .* X[:, j]) )
+            end
+        end
+    end
+
+
+    if npoly >= 3
+        for i in 1:nvar
+            for j in i:nvar
+                for k in j:nvar
+                basis = hcat(basis, (X[:,i] .* X[:, j] .* X[:, k]) )
+            end
+            end
+        end
+    end
+
+
+    if npoly >= 4
+        for i in 1:nvar
+            for j in i:nvar
+                for k in j:nvar
+                    for l in k:nvar
+                basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l]))
+            end
+            end
+        end
+    end
+end
+
+if npoly >= 5
+    for i in 1:nvar
+        for j in i:nvar
+            for k in j:nvar
+                for l in k:nvar
+                    for m in l:nvar
+            basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l] .* X[:,m]))
+        end
+        end
+    end
+end
+end
+end
+
+if npoly >= 6
+    for i in 1:nvar
+        for j in i:nvar
+            for k in j:nvar
+                for l in k:nvar
+                    for m in l:nvar
+                        for n in m:nvar
+            basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l] .* X[:,m] .* X[:,n]))
+        end
+        end
+    end
+end
+end
+end
+end
+
+if ntrig >=1
+    basis = hcat(basis, sin.(X))
+    basis = hcat(basis, cos.(X))
+end
+
+if ntrig >=2
+    basis = hcat(basis, cos.(X) .* cos.(X))
+end
+
+if ntrig >=3
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* sin.(X[:, j]) .* cos.(X[:, j]) ))
+        end
+    end
+end
+
+if ntrig >=4
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* sin.(X[:, j])))
+        end
+    end
+
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* cos.(X[:, j])))
+        end
+    end
+
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (sin.(X[:, i]) .* cos.(X[:, j]) ))
+        end
+    end
+end
+
+    pin = size(basis,2)
+
+    basis_noderivative = deepcopy(basis[:, 1:pin])
+
+    basis_deep = zeros(size(basis,1), 2*size(basis,2))
+
+    basis_deep[:, 1:size(basis,2)] = basis_noderivative
+
+    for k=1:pin
+        basis_deep[:,pin + k] = Y[:, w].* basis_noderivative[:, k]
+    end
+
+    Right_train = zeros(size(basis_deep, 1), size(basis_deep, 2)-1)
+
+    for i = 1:size(basis_noderivative,2)
+        Right_train[:, i] = basis_deep[:,i]
+    end
+
+    for i = size(basis_noderivative,2)+1:size(basis_deep, 2)-1
+        Right_train[:, i] = basis_deep[:,i+1]
+    end
+
+    Left_train = Y[:, w]

```

Comment:
Here is an important part missing / wrongly implemented.

The connected [paper](https://arxiv.org/abs/2004.02322) states on page 2 ( between Eq. 2 and Eq. 3 ) something really important. Normally, you try to solve for a sparse vector in the nullspace of Theta(X, Y) = 0 . But this is really difficult. So what you do is you start to try out different functions of X and Y for the left hand side. This is given in Eq. 3. 

This is what I mean when I say "brute force implementation". Since you are trying out every function in your candidate library, you end up iterating. So think about something like ( pseudocode )

+ Iterate over all functions ( I think here columns ) of Theta which can become the left hand side
+ Solve the sparse regression for column = (all Theta columns except the one on the left hand side ) * Epsilon 
+ Find the best result

This can be done differently as an optimisation problem ( which would not be brute force ).

The reason all your test systems pass is that you can isolate y on the left hand side ( so e.g. the result of the reaction kinetics is something like y = stuff + constant*y ). If you try out the variation I mentioned 

```
jx = 0.6
Vm = 1.5
Km = 0.3
func(u,p,t) = jx/u - Vm
u0 = 0.9
tspan = (0.0,3.0)
dt = 0.01
prob = ODEProblem(func,u0,tspan)
sol = solve(prob, Tsit5(), reltol=1e-8, abstol=1e-8, saveat = dt)
plot(sol)
X = sol.u
DX = map(xi->func(xi, [], 0.0), X)
plot(DX)
using Latexify
Polyorder =  1
Trigorder = 0
ODE_Output3 =SINDYPI(X, DX, Polyorder, Trigorder);
```

You cannot do this anymore. The solution is x*dx = jx - Vm*x , so you wold need to try out anyone of the triple x*dx, constant, x to find a valid solution.

Reply:
@RajDandekar this is really the key issue here.